File: go/analyze/rules_naming.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package analyze

import (
	"context"
	"regexp"
	"strings"

	keystonev1alpha1 "github.com/dogukanturhal/keystone/api/v1alpha1"
)

// Phase A2 — identifier hygiene rules.
//
// PostgreSQL allows identifiers up to NAMEDATALEN - 1 = 63 bytes and
// reserves any identifier starting with `pg_` for system use. Both
// limits are easy to run into with tooling-generated names — index
// names in particular can exceed the limit when concatenating table +
// column + suffix.

// -- Rule: max-identifier-length ---------------------------------------

// MaxIdentifierLength flags identifiers inside CREATE / ALTER / INDEX
// / CONSTRAINT declarations that are 60+ characters long. PostgreSQL
// silently truncates at 63 bytes, which is rarely what the author
// wanted — a truncated suffix collides with other identifiers. Default
// threshold is 60 (leaving 3 bytes of headroom).
type MaxIdentifierLength struct {
	// Threshold is the maximum length below which a name is fine.
	// Default 60. Set 63 to match PG exactly.
	Threshold int
}

func (MaxIdentifierLength) ID() string { return "max-identifier-length" }
func (MaxIdentifierLength) Description() string {
	return "PostgreSQL truncates identifiers at 63 bytes; warn at 60+"
}

// Capture identifiers in the common DDL positions.
var reNamedObjects = regexp.MustCompile(`(?i)\b(CREATE\s+(?:UNIQUE\s+)?INDEX(?:\s+CONCURRENTLY)?|CREATE\s+TABLE(?:\s+IF\s+NOT\s+EXISTS)?|CREATE\s+(?:MATERIALIZED\s+)?VIEW|CREATE\s+SEQUENCE|CREATE\s+FUNCTION|ALTER\s+TABLE|CONSTRAINT|REFERENCES)\s+(?:IF\s+NOT\s+EXISTS\s+)?"?([a-zA-Z_][a-zA-Z0-9_]*)"?`)

func (a *MaxIdentifierLength) Check(_ context.Context, m *Migration) ([]Finding, error) {
	threshold := a.Threshold
	if threshold == 0 {
		threshold = 60
	}
	var out []Finding
	for _, f := range m.Files {
		for _, match := range reNamedObjects.FindAllStringSubmatchIndex(f.Body, -1) {
			// match[2]..match[3] = subgroup 1 (keyword), match[4]..match[5] = subgroup 2 (identifier)
			if match[4] < 0 {
				continue
			}
			name := f.Body[match[4]:match[5]]
			if len(name) < threshold {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(lineOfOffset(f.Body, match[4])),
				Message: "identifier " + name + " is " + itoa(len(name)) + " chars; " +
					"PostgreSQL truncates at 63 bytes and a truncated suffix can collide " +
					"with other objects. Shorten to < " + itoa(threshold) + ".",
			})
		}
	}
	return out, nil
}

// -- Rule: no-pg-prefix-identifier -------------------------------------

// NoPgPrefixIdentifier flags user objects named with a `pg_` prefix.
// PostgreSQL reserves that prefix for system catalogues and will
// refuse to create some object kinds with the prefix at all; worse,
// future PG versions may add a system object with the same name and
// break user code that depended on it.
type NoPgPrefixIdentifier struct{}

func (NoPgPrefixIdentifier) ID() string { return "no-pg-prefix-identifier" }
func (NoPgPrefixIdentifier) Description() string {
	return "user objects named `pg_*` collide with system namespace"
}

// Match identifiers starting with pg_ in creation positions (table,
// index, view, function, sequence, constraint, role, schema). Keep
// narrow — we don't want to flag references to existing pg_ tables.
var rePgPrefix = regexp.MustCompile(`(?i)\b(CREATE\s+(?:UNIQUE\s+)?INDEX(?:\s+CONCURRENTLY)?|CREATE\s+TABLE|CREATE\s+(?:MATERIALIZED\s+)?VIEW|CREATE\s+SEQUENCE|CREATE\s+FUNCTION|CREATE\s+ROLE|CREATE\s+SCHEMA|ALTER\s+TABLE|CONSTRAINT)\s+(?:IF\s+NOT\s+EXISTS\s+)?"?(pg_\w+)"?`)

func (a *NoPgPrefixIdentifier) Check(_ context.Context, m *Migration) ([]Finding, error) {
	var out []Finding
	for _, f := range m.Files {
		for _, match := range rePgPrefix.FindAllStringSubmatchIndex(f.Body, -1) {
			if match[4] < 0 {
				continue
			}
			name := f.Body[match[4]:match[5]]
			if !strings.HasPrefix(strings.ToLower(name), "pg_") {
				continue
			}
			out = append(out, Finding{
				Rule:     a.ID(),
				Severity: keystonev1alpha1.LintLevelWarning,
				File:     f.Name,
				Line:     int32(lineOfOffset(f.Body, match[4])),
				Message: "object named " + name + " uses the reserved `pg_` prefix. " +
					"Future PostgreSQL versions may add a system object with the same " +
					"name and collide. Rename to a project-specific prefix.",
			})
		}
	}
	return out, nil
}

File: rust/keystone-sdk/src/analyze/rules_naming.rs
// SPDX-License-Identifier: Apache-2.0

//! Phase A2 — identifier hygiene. Ported from `rules_naming.go`.

use std::sync::LazyLock;

use regex::Regex;

use super::{finding, line_of_offset, Analyzer, Finding, LintLevel, Migration};

macro_rules! re {
    ($name:ident, $pat:expr) => {
        static $name: LazyLock<Regex> = LazyLock::new(|| Regex::new($pat).expect($pat));
    };
}

re!(
    RE_NAMED_OBJECTS,
    r#"(?i)\b(CREATE\s+(?:UNIQUE\s+)?INDEX(?:\s+CONCURRENTLY)?|CREATE\s+TABLE(?:\s+IF\s+NOT\s+EXISTS)?|CREATE\s+(?:MATERIALIZED\s+)?VIEW|CREATE\s+SEQUENCE|CREATE\s+FUNCTION|ALTER\s+TABLE|CONSTRAINT|REFERENCES)\s+(?:IF\s+NOT\s+EXISTS\s+)?"?([a-zA-Z_][a-zA-Z0-9_]*)"?"#
);
re!(
    RE_PG_PREFIX,
    r#"(?i)\b(CREATE\s+(?:UNIQUE\s+)?INDEX(?:\s+CONCURRENTLY)?|CREATE\s+TABLE|CREATE\s+(?:MATERIALIZED\s+)?VIEW|CREATE\s+SEQUENCE|CREATE\s+FUNCTION|CREATE\s+ROLE|CREATE\s+SCHEMA|ALTER\s+TABLE|CONSTRAINT)\s+(?:IF\s+NOT\s+EXISTS\s+)?"?(pg_\w+)"?"#
);

/// max-identifier-length (default threshold 60)
pub struct MaxIdentifierLength;
impl Analyzer for MaxIdentifierLength {
    fn id(&self) -> &'static str {
        "max-identifier-length"
    }
    fn description(&self) -> &'static str {
        "PostgreSQL truncates identifiers at 63 bytes; warn at 60+"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        const THRESHOLD: usize = 60;
        let mut out = Vec::new();
        for f in &m.files {
            for caps in RE_NAMED_OBJECTS.captures_iter(&f.body) {
                let Some(g) = caps.get(2) else {
                    continue;
                };
                let name = g.as_str();
                if name.len() < THRESHOLD {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Warning,
                    &f.name,
                    line_of_offset(&f.body, g.start()),
                    format!(
                        "identifier {name} is {} chars; \
                         PostgreSQL truncates at 63 bytes and a truncated suffix can collide \
                         with other objects. Shorten to < {THRESHOLD}.",
                        name.len()
                    ),
                ));
            }
        }
        out
    }
}

/// no-pg-prefix-identifier
pub struct NoPgPrefixIdentifier;
impl Analyzer for NoPgPrefixIdentifier {
    fn id(&self) -> &'static str {
        "no-pg-prefix-identifier"
    }
    fn description(&self) -> &'static str {
        "user objects named `pg_*` collide with system namespace"
    }
    fn check(&self, m: &Migration) -> Vec<Finding> {
        let mut out = Vec::new();
        for f in &m.files {
            for caps in RE_PG_PREFIX.captures_iter(&f.body) {
                let Some(g) = caps.get(2) else {
                    continue;
                };
                let name = g.as_str();
                if !name.to_lowercase().starts_with("pg_") {
                    continue;
                }
                out.push(finding(
                    self.id(),
                    LintLevel::Warning,
                    &f.name,
                    line_of_offset(&f.body, g.start()),
                    format!(
                        "object named {name} uses the reserved `pg_` prefix. \
                         Future PostgreSQL versions may add a system object with the same \
                         name and collide. Rename to a project-specific prefix."
                    ),
                ));
            }
        }
        out
    }
}

#[cfg(test)]
mod tests {
    use super::super::{FileBody, Migration};
    use super::*;

    fn mig(body: &str) -> Migration {
        Migration {
            files: vec![FileBody {
                name: "001.sql".into(),
                body: body.into(),
            }],
            ..Default::default()
        }
    }

    #[test]
    fn max_identifier_length() {
        let long = "a".repeat(61);
        let f = MaxIdentifierLength.check(&mig(&format!("CREATE TABLE {long} (id int);")));
        assert_eq!(f.len(), 1);
        assert!(f[0].message.contains("61 chars"));
        // 59 chars is fine.
        let ok = "a".repeat(59);
        assert_eq!(
            MaxIdentifierLength
                .check(&mig(&format!("CREATE TABLE {ok} (id int);")))
                .len(),
            0
        );
    }

    #[test]
    fn pg_prefix() {
        assert_eq!(
            NoPgPrefixIdentifier
                .check(&mig("CREATE TABLE pg_secret (id int);"))
                .len(),
            1
        );
        assert_eq!(
            NoPgPrefixIdentifier
                .check(&mig("CREATE TABLE secret (id int);"))
                .len(),
            0
        );
    }
}
